### ISSUE TITLE
Add GUID collision check

### ISSUE THREAD

**Issue by ETLegacyTeam** [etlegacy/etlegacy#1346]
Author: @rmarquis (Spyhawk)
Date: 2019-09-08
Redmine Issue: 1346, https://dev.etlegacy.com/issues/1346

---

Since PB was responsible for giving unique ID and check for collision (dropping players with non unique ID), it's removal from the open source code led to adding an etkey generator in the client.

However, 2.60b clients that have no GUID can still possibly connect to any mod. In Legacy mod, we added a check in 73f57387ee5a1d170cdde9924bec2eebc75f042d to prevent players with no guid to connect (due to issue with Skill Rating, see #1344, and obviously invalid guid make administration harder). It might be wise to prevent this server side too.

* ~~Add GUID check for validity (empty, "unknown", "NO_GUID", wrong format)~~ (done)
* ~~Add check for duplicate GUID on server (beware of multiple clients use for dev mode though!)~~ (done)
* ~~Drop clients with invalid GUID with a nice message ("Go get ET:L client, or use etkey.org")~~ (done)
* ~~If the check is done server side, the relevant mod code could be removed (Legacy mod is not designed to run on the vanilla server, and Skill Rating depends on database code added to ETLegacy).~~ (done)
* Maybe move some other part of the UserInfo string to server check? (f.e. empty name).

**Comment by rmarquis** [comment]
TODO:
* ~~Also add GUID content check (hex string). Currently only the length is checked.~~ (done)

**Comment by Aranud** [comment]
@rmarquis 

> If the check is done server side, the relevant mod code could be removed (Legacy mod is not designed to run on the vanilla server, and Skill Rating depends on database code added to ETLegacy).

I don't understand what should be done. All the GUID check are already done in game mod (server side so). Am I missing something ?

**Comment by rmarquis** [comment]
For Legacy mod, this is fully done (nice addition btw).

The question is: if this worth adding support for all mods? In which case, the check should be done in the server (sv) part of the code. So third party mods that don't have such a check would have it when running on etlded. If this is ever done, the code in Legacy game code would become redundant and would need to be removed.

I don't know how complex or even possible it is, but I personally _don't_ think this is worth it, since I don't really care for these mods anyway, and some of them already have this feature anyway (like NQ). I'd _personally_ close the ticket and consider it done :)

**Comment by Aciz** [comment]
Continuing the discussion from #1769 

> rmarquis commented
> GUID check has been implemented in #1346. All the database stuff kinda relies on it (SR, WolfAdmin and maybe other stuff), so it's basically made mandatory.
>
> I am aware of one edge case we had in the past, which involves a player connecting to a Legacy server directly from another mod, which kinda leaks the old userinfo into the new one as a result (and thus in some case, no GUID). I am not sure if we actually fixed that case. Ryven might remember some details here.

&nbsp;

> Aciz commented
> I think even on legacy mod you can have no GUID, if the server is running with sv_guidCheck 0?

&nbsp;

> rmarquis commented
> You are right, it's not actually "mandatory" but enabled by default. The ETL client ensure the GUID is properly set, but the old 2.60b client does not, and that's the one that has the edge cases afaik.

&nbsp;

> Aciz commented
> And TM main for example runs with sv_guidCheck 0, as well as TM scrim servers, to keep compatibility with 2.60b... I am literally at loss of words how stupid this cvar is, all it does basically is prevent 2.60b clients by default from playing on ETL servers. This check should be in the mod, not in the server, because it's the mod that cares about if you have GUID, not the server.

&nbsp;

> rmarquis commented
> There's some discussion about this server/mod stuff in #1346. This might need to be reversed if it is really problematic.


I think it's seriously time to reconsider the implementation of this, and move this check back to the game code and away from server code.

The fact that it's in server code means that by default, ETL server is not compatible with 2.60b clients.  For other mods that have their own authentication methods (such as NQ), this also means that 2.60b clients are essentially rejected connection for no reason, they do not need etkey on those mods, yet they are rejected connection because they don't have one.

Having this check in game code would still allow legacy mod to require GUID from clients in order for the various systems tied to it work, but not make other mods fail by default for no reason.

I am also not really sure how 2.60b clients even get to servers ever, when this check is enabled. I've tried to connect to my own ETL server running ETJump with 2.60b client when GUID checking is enabled, but it rejects connection even though I have etkey. From what I remember, this is because 2.60b clients by default do not send GUID on challenge, where this check happens. They only send the GUID on challenge if PunkBuster is installed and have `pb_cl_enable 1`. Now when 2.60b client is fully connected to a server, with a valid etkey/GUID, can the server read the GUID then, even if they have no PB installed and enabled? Or are 2.60b clients completely incompatible with this check, if they do not have PB installed and enabled?

**Comment by rmarquis** [comment]
> From what I remember, this is because 2.60b clients by default do not send GUID on challenge, where this check happens. They only send the GUID on challenge if PunkBuster is installed and have pb_cl_enable 1. 

Need to confirm, but I think you're right here. `pb_cl_enable` has to be enabled on 2.60b.

I have no real strong opinion on where the check should be, as I kinda don't really care about making the feature available to other mods as mentioned above.

However, to me, not enforcing GUID check seems like a non sense - on Legacy mod at least. Features rely on it (SR computation of every player gets messed up if some people play without GUID), and worst, the admin suit really relies on GUID to enforce bans. Incidentally, most existing hacks work on 2.60b only. I don't see it much differently than not allowing protocol 82 to be honest.

So, why is that exactly that TM are allowing empty guid on main? Are they simply doing it because some people can't bother to get an etkey? And on *scrims* too? I thought these were the actually competent guys? How come one extra command line is too much for them? What am I missing here?

If the issue is that indeed is too difficult to users to get a key/enable pb, how can we make it easier? There shouldn't be much valid reason to use 2.60b client to play Legacy mod by now, and there is a big fatty warning on the start screen for 2.60b client when loading Legacy mod. But if really necessary, can we somehow make it the instruction more explicit here?

**Comment by Aciz** [comment]
> So, why is that exactly that TM are allowing empty guid on main? Are they simply doing it because some people can't bother to get an etkey? And on _scrims_ too? I thought these were the actually competent guys? How come one extra command line is too much for them? What am I missing here?

You need to actually install PunkBuster in order to set `pb_cl_enable 1`, which is the big problem here, nobody wants to (and should not be forced to) do that in order to play legacy mod on 2.60b, as having PB installed is not a requirement for having a valid GUID.

**Comment by rmarquis** [comment]
Well, I'd kinda argue an empty GUID is *not* a valid GUID :P

As a corollary question, why is ETL client still not used by these people to play Legacy mod, apart from good old inertia? Can we make the ugly fatty 2.60b warning even uglier, fattier to them? Or better, display a list of "killer features" that only ETL client has that should help them decide to move to ETL client?

F.e., I'm looking at #1736 for SVG crosshair, and honestly I see it as the next "relatively easy to implement" feature that should give another big improvement to ETL client over 2.60b, next to the many others that have been already added.

**Comment by Aciz** [comment]
> Well, I'd kinda argue an empty GUID is _not_ a valid GUID :P

See this is the issue I have with the server rejecting connection at challenge. Just because you don't have PB installed and enabled, does not mean that you necessarily have an empty GUID.  2.60b just does not communicate it at that point, hence why I was wondering about this:

> Now when 2.60b client is fully connected to a server, with a valid etkey/GUID, can the server read the GUID then, even if they have no PB installed and enabled? Or are 2.60b clients completely incompatible with this check, if they do not have PB installed and enabled?

If this is not true, and 2.60b clients (with valid etkey/GUID) still do not communicate this data to the server once they are in, I guess 2.60b is then not ever going to be compatible with this check, unless they have PB actually installed and running. But, if they do communicate it after finishing connection, then I see no reason to reject connection by the server before letting the clients in and doing the check then, at which point it should be done by the mod anyway, since the server never should even care about this information.

**Comment by rmarquis** [comment]
Ah, I see what you mean now.

I just checked it to be sure: 2.60b sends "unknown" when pb isn't enabled, and NO_GUID when the etkey is missing. The GUID isn't sent without pb enabled.

However.... pb is installed in any case, it's just a matter of enabling it. The original installer `WolfET.exe` won't accept to install without also installing pb (but doesn't generate an etkey), while the current installer `WolfET_2_60b_custom.exe` on spashdamage.com has an additional option to enable the PB service at install time (and generate an etkey). I haven't checked other platforms, but these should be a very small minority.

I'm kinda even more lost as to why players would refuse to run the `pb_cl_enable 1`. I mean, in addition to refusing to use the ETL client instead...

**Comment by Aciz** [comment]
> I'm kinda even more lost as to why players would refuse to run the `pb_cl_enable 1`. I mean, in addition to refusing to use the ETL client instead...

I don't think the issue is refusing to set that cvar on, the issue is that the cvar is not available unless you have PB installed and running, which no one wants to do, since it's notorious for causing hitches and stutters during gameplay.

Edit: and no, I would say that majority of players who play on 2.60b do not have PB installed. It has not been required for a long time, and I have personally carried my 2.60b install over through OS reinstalls and different PCs since 2010, and I don't even remember the last time I've had PB installed, so I cannot for example enable this cvar at all.

**Comment by rmarquis** [comment]
Still, PB is installed by default. True the service needs to be enabled, and pb sucks, but that's like 2.60b client anyway.

This still doesn't answer the questions:
* Why 2.60 is still used apart from inertia?
* What we might need to fix so ETL client is at least "as good" as the old client?
* What we can do to encourage more to move to ETL client (which doesn't have that PB issue)?

There are downside of having clients without GUID, and I'd kinda like to avoid these.

Also to reiterate: I'm personally not against the idea of moving the check mod side, and I'm more seeing the issue here from a Legacy perspective only (while you consider etjump here), but the idea of allowing empty guid "just because" is kinda dubious to me.

**Comment by Aciz** [comment]
I'm not saying we should allow empty GUID ever, of course not. But I do not think it's the job of the server engine to check for that, as it's the mod that either requires it or doesn't, hence I am strongly for moving it to the game code and let every other mod make sure that they do what is necessary to avoid empty GUID, if it's harmful for them.

I don't think anyone would in their sane mind say that 2.60b is better than ETL client, it has surpassed it a long time ago (although I do miss my viewlog, I think the reason it was removed is really stupid, but that's a discussion for some other time). My guess for the main reason people still play with 2.60b is simply laziness in converting their 2.60b config to be ETL compatible (so get rid of some renderer cvars and re-do them in ETL basically). I've had an idea for a while that it would be cool to write some sort of web app, which could for example be put in our website, that basically converts 2.60b/ETPro configs to ETL/legacy configs by changing `b_` cvars to their eqivalent `cg_` counterparts (if applicable), remove unused/unsafe renderer cvars etc.

While staying compatible with 2.60b sort of sucks sometimes, it also really isn't that big of a deal by doing stuff like this in the mod, rather than engine.

**Comment by rmarquis** [comment]
Yes, I can fully get behind your reasoning here. Let's move the check back to mod side then.

I'd be strongly in favour of eventually forcing the GUID check (in Legacy mod), but it doesn't have to be that radical if it is very problematic now: some sort of always displayed, rather annoying string on the HUD of people that don't have GUID might work in the short/medium term to encourage them to move to ETL client (or enable PB if they don't wish to do so). We could also treat the `NO_GUID` , `unknown` and actually empty guid differently here, to give them proper instruction.

Just some idea to help fight laziness.

Some sort of "compatibility app" would be great. It's another excellent idea to make it easier to move to ETL client. We just need someone to do the job (thank you for volunteering!...  maybe? :D)

**Comment by jackeri** [comment]
Not so sure if the guid check move to the mod was so good in the end. It also limits us from more easily enforcing any registered guids in the future, and now theres also some issues risen with the move (clients not being to able to connect to servers).

**Comment by rmarquis** [comment]
Yes, my thoughts too. I've no idea why that duplicate keys issue happens.
I'm certainly not against moving the code back to server code, but I guess the multi-mod aspect raised by Aciz need some further consideration.

**Comment by ryzyk-krzysiek** [comment]
To reproduce the issue join server and kill client. Normally client would get timed out after a while which would call `SV_DropClient` and `ClientDisconnect`, if you don't join before that everything works as expected really.

The real problem is if you join before you get timed out, `GAME_CLIENT_CONNECT` will reject connection and change `client->state` to `CS_FREE` causing the needed disconnect code to never run, that makes the qagame hold the player ent and cause subsequent kicks untill it gets flushed by for example `map_restart`.

It never happend when check was on server because on reconnect the guid check was skipped.

The only idea I have right now is maybe try something like this:
```
if (!Q_strncmp(cl->pers.cl_guid, cs_guid, MAX_GUID_LENGTH + 1))
{
	if (ent->inuse)
	{
		G_LogPrintf("Forcing disconnect on active client: %i\n", (int)(ent - g_entities));
		ClientDisconnect(ent - g_entities);
	}
	return "Bad GUID: Duplicate etkey.";
}
```

Basically this https://github.com/etlegacy/etlegacy/blob/e4035d4a5c6c481ea91267a2579f6ec8653f36e9/src/game/g_client.c#L2205-L2213

So disconnect hanging player and upon next connection the player should be able to join. This of course would mean that someone spoofing someone elses guid could kick players from server like that also not sure if it would work in 100% of cases.
